pkg/cloudprovider/providers/azure/azure_cache.go

```diff
@@ -19,63 +19,138 @@ package azure
 	}
 }
 
-func cacheKeyFunc(obj interface{}) (string, error) {
-	return obj.(*timedcacheEntry).key, nil
+// expired returns true if the cache is expired.
+func (t *timedCache) expired() bool {
+	return t.lastRefresh.Add(t.ttl).Before(time.Now())
 }
 
-func (t *timedcache) GetOrCreate(key string, createFunc func() interface{}) (interface{}, error) {
-	entry, exists, err := t.store.GetByKey(key)
-	if err != nil {
+// Get gets data by key. It returns nil if data not found.
+func (t *timedCache) Get(key string) (interface{}, error) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if t.expired() {
+		if err := t.refresh(""); err != nil {
+			return nil, err
+		}
+	}
+
+	entry, ok := t.store[key]
+	if ok {
+		return entry, nil
+	}
+
+	// not cache this vm yet, refresh by key
+	if err := t.refresh(key); err != nil {
 		return nil, err
 	}
-	if exists {
-		return (entry.(*timedcacheEntry)).data, nil
+	entry, ok = t.store[key]
+	if ok {
+		return entry, nil
 	}
 
-	t.lock.Lock()
-	defer t.lock.Unlock()
-	entry, exists, err = t.store.GetByKey(key)
+	// Key still not found, set it to nil. This is required to avoid
+	// relisting nonexist objects.
+	t.store[key] = nil
+	return nil, nil
+}
+
+// refresh updates cache by getter or lister. If key is an empty string,
+// then lister is used to update cache.
+// It should be only called under a lock.
+func (t *timedCache) refresh(key string) error {
+	// Refresh cache by getter.
+	if key != "" && t.getter != nil {
+		data, err := t.getter(key)
+		if err != nil {
+			return err
+		}
+
+		t.store[key] = data
+		return nil
+	}
+
+	// Refresh cache by lister.
+	dataList, err := t.lister()
 	if err != nil {
-		return nil, err
+		return err
 	}
-	if exists {
-		return (entry.(*timedcacheEntry)).data, nil
+	t.store = make(map[string]interface{})
+	for key, data := range dataList {
+		t.store[key] = data
 	}
 
-	if createFunc == nil {
-		return nil, nil
+	t.lastRefresh = time.Now()
+	return nil
+}
+
+// Delete removes data from cache by key.
+func (t *timedCache) Delete(key string) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	// mark data as nil to avoid invoking APIs later.
+	t.store[key] = nil
+}
+
+// List gets a list of data from cache.
+func (t *timedCache) List() ([]interface{}, error) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if t.expired() {
+		if err := t.refresh(""); err != nil {
+			return nil, err
+		}
 	}
-	created := createFunc()
-	t.store.Add(&timedcacheEntry{
-		key:  key,
-		data: created,
-	})
-	return created, nil
+
+	result := make([]interface{}, 0)
+	for _, data := range t.store {
+		if data == nil {
+			continue
+		}
+
+		result = append(result, data)
+	}
+	return result, nil
 }
 
-func (t *timedcache) Delete(key string) {
-	_ = t.store.Delete(&timedcacheEntry{
-		key: key,
-	})
+// AddOrUpdate adds new data or updates existing data in the cache.
+// If the data is nil, then it updates cache from getter.
+func (t *timedCache) AddOrUpdate(key string, data interface{}) error {
```

Comment:
Also consider split this function: 'explicitly setting a value' and 'request update the value' should be 2 things.

Reply:
Add and update actually do same thing